Sources/Datadog/Tracing/Utils/URLRequestInterceptor.swift

```diff
@@ -0,0 +1,84 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import OpenTracing
+
+// TODO: RUMM-452 URLSessionHooks has incomplete implementation
+
+internal enum URLSessionHooks {
```

Comment:
Why not making this object-oriented instead of a set of static functions (aka structured programming)?

To me, this could be simply:
```swift
internal class URLSessionTracingAutoInstrumentation {
   private let tracingFeature: TracingFeature
   private let domainWhitelist: Set<URL>

   init(domainWhitelist: Set<URL>) throws {
      guard let tracing = TracingFeature.instance else {
         throw ProgrammerError(...) // centralized error throwing API misuse
      }
      self.tracingFeature = tracing
      // swizzle `URLSession`
   }

   deinit {
      // unswizzle `URLSession`
   }

   // ... object-oriented methods, not static functions
}
```
This will nicely connect with the public API by using the same error-logging convention as we do for other things:
```swift
public class Datadog {
    // ...

   static func enableURLSessionInstrumentation(domainWhitelist whitelist: Set<URL>) {
      do {
         TracingFeature.instance.urlSessionInstrumentation = try URLSessionTracingAutoInstrumentation(whitelist)
      } catch {
         consolePrint("\(error)")
      }
   }
}
```

Reply:
I've never heard of the term structured programming but the aim here is/was having immutable stateless static blocks.
currently, the only captured reference is `Global.sharedTracer`, which is good enough yet not perfect.

if we need `TracingFeature`, yes we will probably need to change this design to a singleton because there can be only one instrumentation instance only as we swizzle methods globally.
but if possible, I'd prefer keeping static blocks rather than singletons.

Note on `deinit { unswizzle() }`, we can never call unswizzle because this will unswizzle 3rd party libraries as well